### DISCUSSION TITLE
Utility for generating UNIQUE IDs

### DISCUSSION THREAD

**Discussion by cernymatej** [nuxt/nuxt#23137] [Ideas]
When making accessible components, it is often necessary to generate truly unique identifiers for `<label />` + `<input>` pairings / checkbox names etc.
There are several ways to tackle this, but none are perfect. Addressing this issue—which I believe to be quite common—doesn't align with the seamless DX that Nuxt typically offers.

Using only a randomly generated ID without additional logic can lead to hydration mismatches. Furthermore, most solutions don't guarantee a truly unique identifier; instead, they rely on the fact that there's an extremely low probability of encountering a duplicate.

React provides a convenient `useId()` [hook](https://react.dev/reference/react/useId), which does the job quite well and should return a truly unique value. (though I am not certain about it)

Given the emphasis on great DX in Nuxt, I feel this issue should be managed at the framework level.

Would it be feasible to offer a global helper for this?

**Comment by logotip4ik** [comment]
Hello! You can use [`vue-bind-once`](https://github.com/danielroe/vue-bind-once) package created by Daniel. It does not provide id itself, but allow's setting id once and then reuse it for hydration, client or server rendering. In [usage section](https://github.com/danielroe/vue-bind-once#usage) you can see how to provide id by yourself with [nanoid](https://www.npmjs.com/package/nanoid), also you can use something like `crypto.randomUUID()` or `(Math.random() * 10000).toString()` to generate ids

**Reply by cernymatej** [reply]
But neither `crypto.randomUUID()`, `(Math.random() * 10000).toString()`  nor nanoid guarantee a unique value. I know the chances are astronomically small but it is possible to get a duplicate. 😕

**Reply by AloisSeckar** [reply]
So what about creating a store containing an array of already generated IDs? And you'll run generating in cycle until the value is not preset yet (which means once most of the time).

**Reply by logotip4ik** [reply]
Well, nothing will guarantee you a truly unique value. But `randomUUID` offers random levels far beyond usual needs. [See stack overflow answer](https://stackoverflow.com/questions/20999792/does-randomuuid-give-a-unique-id/20999821#20999821)

**Reply by cernymatej** [reply]
If the framework implemented some logic for the ID generation, it could guarantee a truly unique value. It could be based on an index and some other variables.

**Comment by localusercamp** [comment]
This will be cool.

Is creating a global increment variable not the solution?

**Comment by cernymatej** [comment]
Glad to see this being worked on in https://github.com/nuxt/nuxt/pull/23368 🤩

**Comment by paleacci** [comment]
After a lot of tinkering, blood sweat and tears I have managed to provide a working solution for creating uuids in Nuxt3. Feel free to use it

```
// composables/useUniqueId.js
export const useUniqueId = () => {
    const uuid = ref('');

    const generateUUID = () => {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = (Math.random() * 16) | 0,
                v = c == 'x' ? r : (r & 0x3) | 0x8;
            return v.toString(16);
        });
    };

    onMounted(() => {
        uuid.value = generateUUID();
    });

    return uuid;
};
```

simple usage
```
const uuid = useUniqueId()
```

enjoy
